DOC-1754: Advanced Typography plugin documentation 2 (#2578)
tinymce/tinymce-docs: modules/ROOT/partials/configuration/advanced-typography.adoc

[[content_langs]]
== `+content_langs+`

This option sets which languages are available for applying language-specific typography rules.

A user can then select text and select from the available languages. That language’s language-specific typography rules are then applied to the selected text.

*Type:* `+Object+`

=== Example: Using `content_langs`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea', // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode} language',

>   language: 'en_US', // This is required to configure the default language used by typography.

Reviewer: See comment about `typography_default_lang`

  content_langs: [
    { title: 'English', code: 'en' },
    { title: 'Spanish', code: 'es' },
    { title: 'German', code: 'de' }
    ],
});
----

=== Currently supported languages

[cols="1,1"]
|===
|Supported language |Language code 

|Belarusian
|be

|Bulgarian
|bg

|Catalan
|ca

|Danish
|da

|Dutch
|nl

|English (UK)
|en-GB

|English (US)
|en-US

|Esperanto
|eo

|Estonian
|et

|Finnish
|fi

|German
|de

|Greek
|el

|Hungarian
|hu

|Irish
|ga

|Italian
|it

|Latvian
|lv

|Norwegian
|no

|Polish
|pl

|Romanian
|ro

|Russian
|ru

|Servian
|sr

|Slovak
|sk

|Slovenian
|sl

|Spanish
|es

|Swedish
|sv

|Turkish
|tr

|Ukraninan
|uk

|===

[[typography_custom_rules]]

> == `+typography_custom_rules+`

Reviewer: Do we need to make it clear how these interact with other built-in rules? If they are given the same name, do they clobber the originals?

Reviewer 2: `typography_custom_rules` feature was not implemented in this release.

Author: @yacodes,

> `typography_custom_rules` feature was not implemented in this release.

Adding this comment out of an abundance of caution.

I’m parsing the above to mean *the `typography_custom_rules` option is not part of the current Advanced Typography plugin release and, therefore, does not currently need to be documented.*

Given that reading, I quote from the commit message:

> Placed `typography_custom_rules` section inside a comment block. As per @yacode’s note that
>
>> `typography_custom_rules` feature was not implemented in this release.

> // What this option does, precisely.

Reviewer: - [ ] does this need to be resolved before merging?

Author: Marked this as resolved as per [this comment](https://github.com/tinymce/tinymce-docs/pull/2578/files#r1041514007).

  toolbar: '{plugincode}',
  language: 'en_US', // This is required to configure the default language used by typography.
  typography_custom_rules: {
    'en/quotes': (text) => text.replace(/"/g, "'"),
  },
});
----

[[typography_ignore]]
== `+typography_ignore+`

This option tells the {pluginname} plugin to ignore content within specified HTML tags or within specified HTML tags that have specified attributes.

If an HTML tag or an HTML tag with a specified attribute is added to the `typography_ignore` array, the Typograhy plugin’s rules are not applied to material set within that tag.

*Type:* `+Array+`

*Possible values:* `[ '<html-selector>', 'span["<html-selector>"]' ]`

*Default value:* none

=== Example: Using `typography_ignore`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: '{plugincode}',
  toolbar: 'typography',
  language: 'en_US', // This is required to configure the default language used by typography.
  typography_ignore: [ 'code', 'span["lang"]' ]
});
----

[[typography_rules]]
== `+typography_rules+`

This option specifies which rules are available to the {pluginname} plugin.

> The {pluginname} plugin rules are sourced from the https://github.com/typograf/typograf[typograf] library.

Reviewer: Should we make this link an asciidoc attribute?

> [[content_langs]]

Reviewer 3: Missing documentation for `typography_langs` and `typography_default_lang`. Note that `content_langs` already has a [dedicated entry in the docs](https://github.com/tinymce/tinymce-docs/blob/86adeabb818e5270cc2ce19f5dfd31ff67ed30e8/modules/ROOT/partials/configuration/content_langs.adoc)

Reviewer 2: `content_langs` can influence the typography plugin by marking some sections of text with lang attribute. This attribute will be taken into account when typography plugin is about to apply language-specific rules. Not sure if that was the intention here

[[typography_default_lang]]
== `+typography_default_lang+`

> This option gives the user the ability to pre-set a default typography language within the {productname} editor. By setting up the `typography_default_lang`, when the user opens the {productname} editor for the first time, it will automatically default to the new default language applied. If this default language is subesequently changed by the user, the new default will be saved and persistent between documents.

Reviewer 4: ```suggestion
This option gives the user the ability to pre-set a default typography language within the {productname} editor. By setting up the `typography_default_lang`, when the user opens the {productname} editor for the first time, it will automatically default to the new default language applied. If this default language is subsequently changed by the user, the new default will be saved and persistent between documents.
```

By setting up the `typography_default_lang`, each time the user opens the {productname} editor, it will automatically default to the new default language applied.

=== Example: Using `typography_default_langs`

[source,js,subs="attributes+"]
----
tinymce.init({

>   selector: "textarea",  // change this value according to your html
>   menu: { tools: { title: 'Tools', items: 'code wordcount typography' }},
>   plugins: "typography",
>   toolbar: "typography",
>   typography_default_lang: "en-US", // Define the default language here, example: "en-US"

Reviewer 5: ```suggestion
  selector: 'textarea',  // change this value according to your html
  menu: { tools: { title: 'Tools', items: 'code wordcount typography' }},
  plugins: 'typography',
  toolbar: 'typography',
  typography_default_lang: 'en-US', // Define the default language here, example: 'en-US'
```

});
----

[[typography_ignore]]
== `+typography_ignore+`

This option tells the {pluginname} plugin to ignore content within specified HTML tags or within specified HTML tags that have specified attributes.

> If an HTML tag or an HTML tag with a specified attribute is added to the `typography_ignore` array, the Typograhy plugin’s rules are not applied to material set within that tag.

Reviewer 4: ```suggestion
If an HTML tag or an HTML tag with a specified attribute is added to the `typography_ignore` array, the Typography plugin’s rules are not applied to material set within that tag.
```

*Type:* `+Array+`

*Possible values:* `[ '<html-selector>', 'span["<html-selector>"]' ]`

*Default value:* none

=== Example: Using `typography_ignore`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: '{plugincode}',
  toolbar: 'typography',

>   typography_default_lang: [ "en-US", "de", "es" ], // Required to set specific typography language rules.

Reviewer 5: ```suggestion
  typography_default_lang: 'en-US', 
```

Reviewer 5: According to the typography code, this option only accepts a string as its value. Also, it is given an array of strings which is incorrect in other examples,

});
----
////

[[typography_langs]]
== `+typography_langs+`

This option sets which languages are available for applying language-specific typography rules.

A user can then select text and select from the available languages. That language’s language-specific typography rules are then applied to the selected text.

*Type:* `+Object+`

=== Example: Using `typography_langs`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea', // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode} language',
  typography_default_lang: 'en-US',
  typography_langs: [
    'en-US',
    'es',
    'de'
  ],
});
----

=== Currently supported languages

[cols="1,1"]
|===
|Supported language |Language code 

|Belarusian
|be

|Bulgarian
|bg

|Catalan
|ca

|Czech
|cs

|Danish
|da

|Dutch
|nl

|English (UK)
|en-GB

|English (US)
|en-US

|Esperanto
|eo

|Estonian
|et

|Finnish
|fi

|German
|de

|Greek
|el

|Hungarian
|hu

|Irish
|ga

|Italian
|it

|Latvian
|lv

|Norwegian
|no

|Polish
|pl

|Romanian
|ro

|Russian
|ru

|Servian
|sr

|Slovak
|sk

|Slovenian
|sl

|Spanish
|es

|Swedish
|sv

|Turkish
|tr

> |Ukraninan

Reviewer 4: ```suggestion
|Ukrainian
```

[[content_langs]]

> == `+content_langs+`

Reviewer: I don't think that we should document this in typography as it's not typography specific.

This option sets which languages are available for applying language-specific typography rules.

A user can then select text and select from the available languages. That language’s language-specific typography rules are then applied to the selected text.

*Type:* `+Object+`

=== Example: Using `content_langs`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea', // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode} language',

>   typography_default_lang: [ "en-US", "de", "es" ], // Required to set specific typography language rules.
>   content_langs: [
>     { title: 'English', code: 'en' },
>     { title: 'Spanish', code: 'es' },
>     { title: 'German', code: 'de' }
>     ],

Reviewer 5: ```suggestion
  typography_default_lang: 'en-US',
  typography_langs: [
    'en-US',
    'es',
    'de'
  ],
```

});
----

=== Currently supported languages

[cols="1,1"]
|===
|Supported language |Language code 

|Belarusian
|be

|Bulgarian
|bg

|Catalan
|ca

|Czech
|cs

|Danish
|da

|Dutch
|nl

|English (UK)
|en-GB

|English (US)
|en-US

|Esperanto
|eo

|Estonian
|et

|Finnish
|fi

|German
|de

|Greek
|el

|Hungarian
|hu

|Irish
|ga

|Italian
|it

|Latvian
|lv

|Norwegian
|no

|Polish
|pl

|Romanian
|ro

|Russian
|ru

|Servian
|sr

|Slovak
|sk

|Slovenian
|sl

|Spanish
|es

|Swedish
|sv

|Turkish
|tr

|Ukraninan
|uk

|===

////

> [[typography_custom_rules]]

Reviewer: - [ ] remove all code that isn't ready.

Reviewer 6: @ephox-mogran Removed typography_custom_rules template for now.
@brian-forte ref only

== `+typography_custom_rules+`

// What this option does, precisely.
// questions still about `typography_custom_rules`.
// from the supplied example it is unclear (to me, at least) what it does.
// is `/"/g, "'"` just `sed` and `regex`?
// and, even if so, I remain unclear as to what
//     'en/quotes': (text) => text.replace(/"/g, "'"),
// does to
//     (text)
// other than replace ? it with
//     text.replace(/"/g, "'")

*Type:* `+Object+`

=== Example: Using `typography_custom_rules`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode}',

>   typography_default_lang: [ "en-US", "de", "es" ], // Required to set specific typography language rules.

Reviewer 5: ```suggestion
  typography_default_lang: "en-US",
```

  typography_custom_rules: {
    'en/quotes': (text) => text.replace(/"/g, "'"),
  },
});
----
////

[[typography_default_lang]]
== `+typography_default_lang+`

This option gives the user the ability to pre-set a default typography language within the {productname} editor. By setting up the `typography_default_lang`, when the user opens the {productname} editor for the first time, it will automatically default to the new default language applied. If this default language is subesequently changed by the user, the new default will be saved and persistent between documents.

>  By setting up the `typography_default_lang`, each time the user opens the {productname} editor, it will automatically default to the new default language applied.

Reviewer: Leading space.

Reviewer 6: Resolved

=== Example: Using `typography_langs`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: "textarea",  // change this value according to your html
  plugins: "typography",
  toolbar: "typography",
  typography_default_lang: [ "en-US", "de", "es" ], // Required to set specific typography language rules.
});
----

[[typography_rules]]
== `+typography_rules+`

This option specifies which rules are available to the {pluginname} plugin.

The {pluginname} plugin rules are sourced from the https://github.com/typograf/typograf[typograf] library.

All rules supported by the {pluginname} plugin are documented in the https://github.com/typograf/typograf/blob/dev/docs/RULES.en-US.md[Rules of typograf].

*Type:* `+Array+`

=== Example: Using `typography_rules`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode}',

>   language: 'en_US', // This is required to configure the default language used by typography.

Reviewer 5: ```suggestion
  typography_default_lang: 'en_US', // This is required to configure the default language used by typography.
```

*Type:* `+Object+`

=== Example: Using `content_langs`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea', // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode} language',
  language: 'en_US', // This is required to configure the default language used by typography.
  content_langs: [
    { title: 'English', code: 'en' },
    { title: 'Spanish', code: 'es' },
    { title: 'German', code: 'de' }
    ],
});
----

=== Currently supported languages

[cols="1,1"]
|===
|Supported language |Language code 

|Belarusian
|be

|Bulgarian
|bg

|Catalan
|ca

|Czech
|cs

|Danish
|da

|Dutch
|nl

|English (UK)
|en-GB

|English (US)
|en-US

|Esperanto
|eo

|Estonian
|et

|Finnish
|fi

|German
|de

|Greek
|el

|Hungarian
|hu

|Irish
|ga

|Italian
|it

|Latvian
|lv

|Norwegian
|no

|Polish
|pl

|Romanian
|ro

|Russian
|ru

|Servian
|sr

|Slovak
|sk

|Slovenian
|sl

|Spanish
|es

|Swedish
|sv

|Turkish
|tr

|Ukraninan
|uk

|===

// typography_default_lang
// @bforte review required
[[typography_default_lang]]
== `+typography_default_lang+`

> This option gives the user the ability to pre-set a default typography language within the {productname} editor.  By setting up the `typography_default_lang`, each time the user opens the {productname} editor, it will automatically default to the new default language applied.

Reviewer 2: If the user has selected another language before — it will persist between the documents, so this is incorrect.

_By setting up the `typography_default_lang`, if the user opens the {productname} editor for the first time, it will automatically default to the new default language applied. If the language was changed by the user, then it will be saved and persistent between the documents._

Author: Replaced introductory `typography_default_lang` paragraph with (lightly edited) re-write by @yacodes.

*Default value:* none

=== Example: Using `typography_ignore`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: '{plugincode}',
  toolbar: 'typography',
  language: 'en_US', // This is required to configure the default language used by typography.
  typography_ignore: [ 'code', 'span["lang"]' ]
});
----

[[typography_rules]]
== `+typography_rules+`

This option specifies which rules are available to the {pluginname} plugin.

The {pluginname} plugin rules are sourced from the https://github.com/typograf/typograf[typograf] library.

All rules supported by the {pluginname} plugin are documented in the https://github.com/typograf/typograf/blob/dev/docs/RULES.en-US.md[Rules of typograf].

*Type:* `+Array+`

=== Example: Using `typography_rules`

[source,js,subs="attributes+"]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: '{plugincode}',
  toolbar: '{plugincode}',
  language: 'en_US', // This is required to configure the default language used by typography.
  typography_rules: {
    en: ['en/quotes', 'en/spaces']
    es: ['es/quotes', 'es/spaces']

>     }

Reviewer 2: Should be an array of rules, we do not support language-specific rules, yet.
```js
  typography_rules: [
    'common/punctuation/quote',
    'en-US/dash/main',
    'common/nbsp/afterParagraphMark',
    'common/nbsp/afterSectionMark',
  ]
```

Author: To quote the commit message:

> Changed section of `typography_rules` example code to array, as per @yacodes note.
